src/qibo/models/tsp.py

```diff
@@ -0,0 +1,172 @@
+from qibo.symbols import X, Y, Z
+from qibo.models import Circuit, QAOA
+from qibo.core.hamiltonians import Hamiltonian, SymbolicHamiltonian
+from qibo import gates
+import numpy as np
+from collections import defaultdict
+
+
+def calculate_two_to_one(num_cities):
+    """
+    Args:
+        num_cities: number of cities
+
+    Returns:
+        a matrix to faciliate conversion from 2 dimension to one dimension
+    """
+    return np.arange(num_cities ** 2).reshape(num_cities, num_cities)
+
+
+def tsp_phaser(distance_matrix, dense=True):
+    """
+
+    Args:
+        distance_matrix: a numpy matrix encoding the distances
+        dense: whether the hamiltonian is dense,
+
+    Returns: a hamiltonian representing the TSP(traveling salesman problem) phaser according to
+    `arxiv:1709.03489<https://arxiv.org/abs/1709.03489>` by Hadfield (2017).
+
+    """
+    num_cities = distance_matrix.shape[0]
+    two_to_one = calculate_two_to_one(num_cities)
+    form = 0
+    for i in range(num_cities):
+        for u in range(num_cities):
+            for v in range(num_cities):
+                if u != v:
+                    form += distance_matrix[u, v] * Z(int(two_to_one[u, i]))* Z(
+                        int(two_to_one[v, (i + 1) % num_cities]))
+    ham = SymbolicHamiltonian(form)
+    if dense:
+        ham = ham.dense
+    return ham
+
+
+def tsp_mixer(num_cities, dense=True):
+    """
+    Args:
+        num_cities: number of cities in TSP
+        dense: whether the hamiltonian is dense
+
+    Returns: a mixer hamiltonian representing the TSP(traveling salesman problem) phaser
+    according to
+    `arxiv:1709.03489<https://arxiv.org/abs/1709.03489>` by Hadfield (2017).
+
+    """
+    two_to_one = calculate_two_to_one(num_cities)
+    splus = lambda u, i: X(int(two_to_one[u, i])) + 1j * Y(int(two_to_one[u, i]))
+    sminus = lambda u, i: X(int(two_to_one[u, i])) - 1j * Y(int(two_to_one[u, i]))
+    form = 0
+    for i in range(num_cities):
+        for u in range(num_cities):
+            for v in range(num_cities):
+                if u != v:
+                    form += splus(u, i) * splus(v, (i + 1) % num_cities) * sminus(u, (
+                            i + 1) % num_cities) * sminus(v, i) + sminus(u, i) * sminus(v, (
+                            i + 1) % num_cities) * splus(u, (i + 1) % num_cities) * splus(
+                        v, i)
+    ham = SymbolicHamiltonian(form)
+    if dense:
+        ham = ham.dense
+    return ham
+
+
+class tsp:
```

Comment:
Following my comment above, if we could reuse the TSP hamiltonian here, would be great!